File: session-06-capstone/go/kitchen/service.go
package kitchen

import (
	"fmt"
	"math/rand"
	"time"
)

type Preparation struct {
	OrderID int    `json:"orderId"`
	Drink   string `json:"drink"`
	Status  string `json:"status"`
}

type KitchenService interface {
	Prepare(orderID int, drink string) (*Preparation, error)
}

type RealKitchenService struct{}

func (s *RealKitchenService) Prepare(orderID int, drink string) (*Preparation, error) {
	return &Preparation{OrderID: orderID, Drink: drink, Status: "preparing"}, nil
}

type ChaosKitchenService struct{}

func (s *ChaosKitchenService) Prepare(orderID int, drink string) (*Preparation, error) {
	roll := rand.Float64()

	if roll < 0.3 {
		return nil, fmt.Errorf("kitchen equipment malfunction! Order #%d failed", orderID)
	}

	if roll < 0.5 {
		delay := time.Duration(2000+rand.Intn(4000)) * time.Millisecond
		fmt.Printf("[Chaos] Injecting %s latency for order #%d\n", delay, orderID)
		time.Sleep(delay)
	}

	if roll < 0.6 {
		fmt.Printf("[Chaos] Returning bad data for order #%d\n", orderID)
		return &Preparation{OrderID: orderID, Drink: drink, Status: "UNKNOWN_STATUS"}, nil
	}

	return &Preparation{OrderID: orderID, Drink: drink, Status: "preparing"}, nil
}

File: session-06-capstone/python/kitchen/service.py
import random
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass
class Preparation:
    order_id: int
    drink: str
    status: str


class KitchenService(ABC):
    @abstractmethod
    def prepare(self, order_id: int, drink: str) -> Preparation:
        ...


class RealKitchenService(KitchenService):
    def prepare(self, order_id: int, drink: str) -> Preparation:
        return Preparation(order_id=order_id, drink=drink, status="preparing")


class ChaosKitchenService(KitchenService):
    def prepare(self, order_id: int, drink: str) -> Preparation:
        roll = random.random()

        if roll < 0.3:
            raise RuntimeError(f"Kitchen equipment malfunction! Order #{order_id} failed.")

        if roll < 0.5:
            delay = random.uniform(2.0, 6.0)
            print(f"[Chaos] Injecting {delay:.1f}s latency for order #{order_id}")
            time.sleep(delay)

        if roll < 0.6:
            print(f"[Chaos] Returning bad data for order #{order_id}")
            return Preparation(order_id=order_id, drink=drink, status="UNKNOWN_STATUS")

        return Preparation(order_id=order_id, drink=drink, status="preparing")
